### PR TITLE
MDNS hostname match DHCP. Fix collision mangling

### LINKED ISSUE #6869: Web Workflow: network mDNS hostname often differs from local-device mDNS hostname (links & redirects don't work)
### CircuitPython version

```python
Adafruit CircuitPython 8.0.0-beta.0 on 2022-08-18; Adafruit QT Py ESP32S2 with ESP32S2
```

### Code/REPL

```python
# From another device on the network, "augmented" names are found:
m = mdns.Server(wifi.radio)
for service in m.find(service_type="_circuitpython", protocol="_tcp", timeout=5):
        print(service.instance_name)
        print(service.hostname)
        print(f"{service.ipv4_address}:{service.port}")
"""
Adafruit QT Py ESP32S2-5
cpy-f66961-281
192.168.6.180:80
"""

# So these work:
# curl -v -L http://192.168.6.180/cp/devices.json
# curl -v -L http://cpy-f66961-281/cp/devices.json

# But the device itself does not seem to know about the augmented name:
print(m.hostname)
# cpy-f66961
print(m.instance_name)
# cpy-f66961
# NOTE: this is completely different from the Instance Name reported from other devices using .find

# So this does not work:
# curl -v -L http://cpy-f66961.local/cp/devices.json
```

### Behavior

From the network perspective, CircuitPython mDNS hostnames often have a hyphen and integer appended to the hostname. CircuitPython mDNS `server.find` [code](https://gist.github.com/anecdata/06045da4eb109e7467d9e8068a8dffad) will expose these augmented hostnames (and instance names), and browsers and curl will find the devices using these augmented names.

Anyone who has encountered things like `iPhone (3)` has encountered a related issue, I suspect.

However, the device itself [currently] does not recognize itself with these augmented names. The Web API will report something like:

**Hostname:**
[cpy-f66961](http://cpy-f66961.local/)
**IP:**
[192.168.6.180](http://192.168.6.180/)

and clicking the hostname link on the Web API will fail.

Redirects also fail since the device doesn't answer to its nominal hostname:
```
% curl -v -L http://circuitpython.local/cp/devices.json
*   Trying 192.168.6.180:80...
* Connected to circuitpython.local (192.168.6.180) port 80 (#0)
> GET /cp/devices.json HTTP/1.1
> Host: circuitpython.local
> User-Agent: curl/7.79.1
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 307 Temporary Redirect
< Connection: close
< Content-Length: 0
< Location: http://cpy-f66961.local/cp/devices.json
< Access-Control-Allow-Credentials: true
< Vary: Origin, Accept, Upgrade
< Access-Control-Allow-Origin: 
< 
* Closing connection 0
* Issue another request to this URL: 'http://cpy-f66961.local/cp/devices.json'
* Could not resolve host: cpy-f66961.local
* Closing connection 1
curl: (6) Could not resolve host: cpy-f66961.local
```
or simply:
```
% curl -v -L http://cpy-f66961.local/cp/devices.json 
* Could not resolve host: cpy-f66961.local
* Closing connection 0
curl: (6) Could not resolve host: cpy-f66961.local
```

### Description

When a device mDNS hostname has been augmented:
1. the "Hostname:" link on the Web API web page will fail
2. the [File REST API ](https://docs.circuitpython.org/en/latest/docs/workflows.html#web)breaks

### Additional information

Since the device knows its own IPv4 address, perhaps one solution could be to redirect to the IPv4 address instead of to the network-invalid mDNS local-device hostname, and not redirect at all if the entry URL is the IP address.

### PR DESCRIPTION
**Pull request by tannewt**
MDNS defaults to the hostname used by DHCP/LWIP since it is now unique. This makes it the same either way.

This also updates esp-protocols (used for mdns) with a patch to ensure that mangled names are the ones that collide. (circuitpython.local collides but was causing cpy- to be mangled.)

Fixes #6869 again.

@retiredwizard this changes the hostname slightly because it hit an assert when over 32 characters. cpy, board and mac are now dash separated too. That way the board's _ don't confusing things.

@anecdata Please test this for the MDNS collision mangling. I just checked that MDNS still works.

### CODE REVIEW

**Review by dhalbert — DISMISSED** [review]
This looks fine to me but we can await testing by @anecdata and @RetiredWizard as well.

**Comment by tannewt** [comment]
It should be fine in a beta too. I tested that MDNS still works.

**Comment by dhalbert** [comment]
Merging and can be tested in the beta.

**Comment by anecdata** [comment]
I'm getting the same hostname / mdns hostname for all four devices I set up: `cpy-t_qtpy_esp32s2-3ffe2e90`, despite unique MAC addresses.

File: ports/espressif/common-hal/wifi/__init__.c [hunk 1]
```diff
@@ -199,11 +199,20 @@ void common_hal_wifi_init(bool user_initiated) {
...
+    // Avoid double _ in the hostname.
+    if (CIRCUITPY_BOARD_ID[board_trim] == '_') {
+        board_trim++;
+    }
+
+    char cpy_default_hostname[board_len + (MAC_ADDRESS_LENGTH * 2) + 6];
     uint8_t mac[MAC_ADDRESS_LENGTH];
     esp_wifi_get_mac(ESP_IF_WIFI_STA, mac);
-    sprintf(cpy_default_hostname, "cpy_%s_%x", CIRCUITPY_BOARD_ID, (unsigned int)mac);
+    snprintf(cpy_default_hostname, sizeof(cpy_default_hostname), "cpy-%s-%x", CIRCUITPY_BOARD_ID + board_trim, (unsigned int)mac);
```
Comment (RetiredWizard) [inline ports/espressif/common-hal/wifi/__init__.c:215]: My original code to convert the MAC address doesn't work, we should use something closer to your original code in server.c:
```py
snprintf(cpy_default_hostname, sizeof(cpy_default_hostname), "cpy-%s-%02x%02x%02x%02x%02x%02x", CIRCUITPY_BOARD_ID + board_trim, mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
```

**Review by RetiredWizard — CHANGES_REQUESTED** [review]

File: ports/espressif/common-hal/wifi/__init__.c (hunk 1 above)
Comment (anecdata) [reply ports/espressif/common-hal/wifi/__init__.c:215]: I wondered where that `t` came from... it's the end of `Adafruit_...`. Would be nice to split it at a `_`, but probably not worth the code.

**Comment by RetiredWizard** [comment]
Tested on BPI PIcow S3 and Hacktablet, the Web workflow page and my DHCP server both show the expected hostname with the correct :blush: MAC address.

File: ports/espressif/common-hal/wifi/__init__.c (hunk 1 above)
Comment (tannewt) [reply ports/espressif/common-hal/wifi/__init__.c]: I looked at going to the next _ but in some cases you do lose something helpful. In some cases the remainder is the chip number (such as 6 from esp32c6) for example.

**Comment by anecdata** [comment]
Looks good, no mangling observed. mdns hostname correctly shows each MAC address. WW pages load with curl or browser by IP, mdns hostname, or `circuitpython.local[.]`

Links to other devices on the WW pages are correct.

File: ports/espressif/common-hal/wifi/__init__.c (hunk 1 above)
Comment (anecdata) [reply ports/espressif/common-hal/wifi/__init__.c]: could drop to 3 octets of MAC, to free up more of the board name

Comment (tannewt) [reply ports/espressif/common-hal/wifi/__init__.c]: I looked what dropping one octet would do and it didn't seem too helpful. I thought about xoring the three octets together to ensure all bits are included.

**Review by dhalbert — APPROVED** [review]
Thanks @anecdata and @RetiredWizard for testing!

There is some build problem with the latex pdf doc build. I am going to ignore that for now and debug it separately. I will go ahead and merge this.

**Pull request merged**